File: WebControls/AccelaMultipleControl.js
/**
* <pre>
*
*  Accela Citizen Access
*  File: AccelaMultipleControl.js
*
*  Accela, Inc.
*  Copyright (C): 2014
*
*  Description: Client scripts for AccelaMultipleControl control.
*
*  Notes:
* $Id: AccelaMultipleControl.js
*  Revision History
*  Date,            Who,        What
*  March 06, 2014   Ian Chen     Initial.
* </pre>
*/

var AccelaMultipleControl = new function () {
    this.validatorForRequired = validatorForRequired;
    this.validatorForDuplicate = validatorForDuplicate;
    this.initializeControl = initializeControl;
    this.adjustSubFieldLayout = adjustSubFieldLayout;

    function initializeControl(currentParentId, nextParentId, controlWidth) {
        this.controlWidth = controlWidth;

        if (typeof (Sys) != "undefined" && Sys.WebForms && Sys.WebForms.PageRequestManager && Sys.WebForms.PageRequestManager.getInstance) {
            with (Sys.WebForms.PageRequestManager.getInstance()) {
                add_pageLoaded(function() {
                    setSubControlTabOrder(currentParentId, nextParentId);
                });
            }
        }
    }

    /*
    * Func: ValidatorForRequired
    * Desc: To validate required of the multiple control.
    * Para: source  
    *       args  
    */
    function validatorForRequired(source, args) {
        var parentId = source.ValidatorCallbackBehavior._callbackControlID;
        var validChildIds = new Array();
        var invalidChildIds = new Array();

        $(AccelaCompositeControl.getChildControl(parentId)).each(function () {
            var childId = $(this).attr('id');
            var childValue = AccelaCompositeControl.getChildControlValue(childId);

            if (!isNullOrEmpty(childValue)) {
                validChildIds[validChildIds.length] = childId;
            } else {
                invalidChildIds[invalidChildIds.length] = childId;
            }
        });

        // these propeties will be used to validation, for ValidatorCallbackBehavior.js use. 
        source.isCompositeControl = true;
        source.validateType = 'required';
        source.validChildIds = validChildIds;
        source.invalidChildIds = invalidChildIds;

        args.IsValid = (invalidChildIds.length == 0);
        return args.IsValid;
    }

    /*
    * Func: ValidatorForDuplicate
    * Desc: To validate duplicate of the multiple control.
    * Para: source  
    *       args  
    */
    function validatorForDuplicate(source, args) {
        var parentId = source.ValidatorCallbackBehavior._callbackControlID;
        var validChildIds = new Array();
        var invalidChildIds = new Array();

        var duplicateObjs = new Array();
        $(AccelaCompositeControl.getChildControl(parentId)).each(function () {
            var childId = $(this).attr('id');
            var childValue = AccelaCompositeControl.getChildControlValue(childId);

            var isContainValue = false;
            $.each(duplicateObjs, function () {
                var thisValue = source.ValidatorCallbackBehavior._validationIgnoreCase ? childValue.toLowerCase() : childValue;
                var otherValue = source.ValidatorCallbackBehavior._validationIgnoreCase ? this.childValue.toLowerCase() : this.childValue;

                if (thisValue.trim() == otherValue.trim()) {
                    this.childIds[this.childIds.length] = childId;

                    isContainValue = true;
                    return false;
                }
            });

            if (!isContainValue) {
                var childIds = new Array();
                childIds[0] = childId;
                duplicateObjs[duplicateObjs.length] = { childValue: childValue, childIds: childIds };
            }
        });

        var foundInvalid = false;
        $.each(duplicateObjs, function() {
            if (foundInvalid || isNullOrEmpty(this.childValue) || this.childIds.length == 1) {
                validChildIds = validChildIds.concat(this.childIds);
            } else if (!foundInvalid && this.childIds.length > 1) {
                invalidChildIds = this.childIds;
                foundInvalid = true;
            }
        });

        // these propeties will be used to validation, for ValidatorCallbackBehavior.js use. 
        source.isCompositeControl = true;
        source.validateType = 'duplicate';
        source.validChildIds = validChildIds;
        source.invalidChildIds = invalidChildIds;

        args.IsValid = (invalidChildIds.length == 0);
        return args.IsValid;
    }

    /*
    * Func: setSubControlTabOrder
    * Desc: Set the sub control's tab order between two related multiple control.
    * Para: currentParentId  
    *       nextParentId  
    */
    function setSubControlTabOrder(currentParentId, nextParentId) {
        if (!currentParentId || !nextParentId) {
            return;
        }

        var childControlCount = AccelaCompositeControl.getChildControl(currentParentId).length;
        if (childControlCount <= 1) {
            return;
        }

        var arrTabOrderId = [];
        var nextHelpId = nextParentId + '_help';

        for (var i = 0; i < childControlCount; i++) {
            var currentChildId = currentParentId + GlobalConst.ChildControlIdPrefix + i;
            var nextChildId = nextParentId + GlobalConst.ChildControlIdPrefix + i;
            arrTabOrderId.push(currentChildId);

            if (i == 0 && $.exists($('#' + nextHelpId))) {
                arrTabOrderId.push(nextHelpId);
            }

            arrTabOrderId.push(nextChildId);
        }

        $('#' + currentParentId + ',#' + nextParentId + ', #' + nextHelpId).keydown(function (e) {
            var keyCode = e.keyCode ? e.keyCode : e.which;
            if (keyCode == 9 || e.shiftKey && keyCode == 9) {
                var obj = e.target || e.srcElement;
                var index = arrTabOrderId.indexOf(obj.id);
                var nextObjId;

                if (e.shiftKey) {
                    if (index - 1 >= 0) {
                        nextObjId = arrTabOrderId[index - 1];
                    }
                } else if (index + 1 < arrTabOrderId.length) {
                    nextObjId = arrTabOrderId[index + 1];
                }

                if (nextObjId) {
                    e.preventDefault();
                    $('#' + nextObjId).focus();
                }
            }
        });
    }

    /*
    * Func: adjustSubFieldLayout
    * Desc: Adjust sub label and sub control layout.
    * Para: parentId  
    *       controlType
    *       layoutType
    */
    function adjustSubFieldLayout(parentId, controlType, layoutType) {
        var parentControl = $('#' + parentId);
        var childrenControlCount = AccelaCompositeControl.getChildControl(parentId).length;
        var tdSubLabels = parentControl.find('td.SubLabel');
        var subControls = parentControl.find('td.SubControl ' + controlType);

        var controlWidthWithUnit = isNullOrEmpty(AccelaMultipleControl.controlWidth) ? parentControl.css('width') : AccelaMultipleControl.controlWidth;
        var parentControlWidth = controlWidthWithUnit.lastIndexOf('em') != -1 ? parseFloat(controlWidthWithUnit) * 10 : parseInt(controlWidthWithUnit);

        var subLabelSpace = 0;
        var subLabelWidth = 0;

        if (childrenControlCount > 1) {
            subLabelSpace = 8;
            subLabelWidth = $(tdSubLabels[childrenControlCount - 1]).width();
        }

        var subControlWidth = parentControlWidth - subLabelWidth - subLabelSpace;

        if (controlType == 'select') {
            if (!isNullOrEmpty(AccelaMultipleControl.controlWidth)) {
                if (isFireFox()) {
                    subControlWidth += 2;
                } else {
                    subControlWidth += 5;
                }
            } else {
                if (childrenControlCount == 1) {
                    if ($.browser.msie) {
                        subControlWidth += 1;
                    } else if (isFireFox()) {
                        subControlWidth += 6;
                    } else {
                        subControlWidth += 4;
                    }
                } else {
                    if (isFireFox()) {
                        subControlWidth += 3;
                    } else if (!$.browser.msie) {
                        subControlWidth += 1;
                    }
                }
            }
        }
        else if (controlType == 'input[type=text]') {
            if (isNullOrEmpty(AccelaMultipleControl.controlWidth)) {
                if (childrenControlCount > 1) {
                    subControlWidth -= 3;
                } else if ($.browser.msie) {
                    subControlWidth -= 2;
                }
            }
        }

        if (childrenControlCount == 1) {
            $('#' + parentId + GlobalConst.ChildControlIdPrefix + '0').width(subControlWidth/10 + 'em');
        } else {
            for (var i = 0; i < childrenControlCount; i++) {
                $(tdSubLabels[i]).width(subLabelWidth + subLabelSpace);
                $(subControls[i]).width(subControlWidth/10 + 'em');
            }
        }

        if (layoutType == 'Vertical') {
            $('#' + parentId + '_help').css(($.global.isRTL ? 'margin-left' : 'margin-right'), '0');
        }
    }
};

File: WebControls/AccelaMultipleControl.cs
#region Header

/**
 * <pre>
 *
 *  Accela Citizen Access
 *  File: AccelaMultipleControl.cs
 *
 *  Accela, Inc.
 *  Copyright (C): 2014
 *
 *  Description: The composite control for multiple controls.
 *
 *  Notes:
 * $Id: AccelaMultipleControl.cs 238998 2012-12-04 09:10:00Z ACHIEVO\alan.hu $.
 *  Revision History
 *  Date,            Who,        What
 *  Dec 4, 2012      Alan Hu     Initial.
 * </pre>
 */

#endregion Header

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web.UI;
using System.Web.UI.WebControls;
using Accela.ACA.Common;

[assembly: WebResource("Accela.Web.Controls.AccelaMultipleControl.js", "text/javascript")]

namespace Accela.Web.Controls
{
    /// <summary>
    /// This class implement the composite control that render the multiple controls which have the same control type.
    /// </summary>
    public class AccelaMultipleControl : AccelaCompositeControl
    {
        #region Properties

        /// <summary>
        /// Gets or sets the child control count
        /// </summary>
        public int ChildControlCount
        {
            get
            {
                return Convert.ToInt32(ViewState["ChildControlCount"]);
            }

            set
            {
                ViewState["ChildControlCount"] = value;
            }
        }

        /// <summary>
        /// Gets or sets the control type.
        /// </summary>
        public ControlType ChildControlType { get; set; }

        /// <summary>
        /// Gets or sets the child control sub label.
        /// </summary>
        public string ChildControlSubLabel { get; set; }

        /// <summary>
        /// Gets or sets the child control sub label full name.
        /// </summary>
        public string ChildControlSubLabelFullName { get; set; }

        /// <summary>
        /// Gets or sets the duplicate validation setting.
        /// </summary>
        public DuplicateValidateSetting DuplicateValidate { get; set; }

        /// <summary>
        /// Gets or sets the AccelaDropDownList's set.
        /// </summary>
        public AccelaDropDownList DropDownListSet { get; set; }

        /// <summary>
        /// Gets or sets the AccelaTextBox's set.
        /// </summary>
        public AccelaTextBox TextBoxSet { get; set; }

        /// <summary>
        /// Gets the child control collection.
        /// </summary>
        public IList<IAccelaControl> Children
        {
            get
            {
                IList<IAccelaControl> children = new List<IAccelaControl>();

                for (int i = 0; i < ChildControlCount; i++)
                {
                    string controlID = string.Format("{0}{1}", CHILD_CONTROL_ID_PREFIX, i);
                    IAccelaControl accelaControl = FindControl(controlID) as IAccelaControl;

                    if (accelaControl != null)
                    {
                        children.Add(accelaControl);
                    }
                }

                return children;
            }
        }

        /// <summary>
        /// Gets or sets next focus control id.
        /// </summary>
        public string NextFocusControlID { get; set; }

        /// <summary>
        /// Gets or sets the control width.
        /// </summary>
        public Unit ControlWidth { get; set; }

        #endregion Properties

        #region Public Methods

        /// <summary>
        /// Initialize extender controls.
        /// </summary>
        public override void InitExtenderControl()
        {
        }

        /// <summary>
        /// Is hide the required indicate.
        /// </summary>
        /// <returns>If need hide the required indicate.</returns>
        public override bool IsHideRequireIndicate()
        {
            return false;
        }

        /// <summary>
        /// Check the fields is required or not.
        /// </summary>
        /// <returns>If the fields is required.</returns>
        public override bool IsRequired()
        {
            return IsFieldRequired;
        }

        /// <summary>
        /// Clear value, it not implement now.
        /// </summary>
        public override void ClearValue()
        {
        }

        /// <summary>
        /// Disable the edit form, it not implement now.
        /// </summary>
        public override void DisableEdit()
        {
        }

        /// <summary>
        /// Enable the edit form, it not implement now.
        /// </summary>
        public override void EnableEdit()
        {
        }

        /// <summary>
        /// Get the multiple control's value, split with split character.
        /// </summary>
        /// <returns>The multiple control's value.</returns>
        public override string GetValue()
        {
            string result = string.Empty;

            foreach (IAccelaControl accelaControl in Children.Where(accelaControl => accelaControl != null))
            {
                string value = accelaControl.GetValue();

                if (!string.IsNullOrEmpty(value))
                {
                    result += value + ACAConstant.SPLIT_CHAR;
                }
            }

            if (!string.IsNullOrEmpty(result))
            {
                result = result.TrimEnd(ACAConstant.SPLIT_CHAR);
            }

            return result;
        }

        /// <summary>
        /// Set the multiple control's value.
        /// </summary>
        /// <param name="value">The value that want to set.</param>
        public override void SetValue(string value)
        {
            if (string.IsNullOrEmpty(value))
            {
                return;
            }

            string[] values = value.Split(ACAConstant.SPLIT_CHAR);

            for (int i = 0; i < values.Length; i++)
            {
                IAccelaControl accelaControl = Children[i];
                accelaControl.SetValue(values[i]);
            }
        }

        #endregion Public Methods

        #region Protected Control Event

        /// <summary>
        /// Create children controls
        /// </summary>
        protected override void CreateChildControls()
        {
            Controls.Clear();

            base.CreateChildControls();

            bool isNeedAddSubLabel = !string.IsNullOrEmpty(ChildControlSubLabel) && ChildControlCount > 1;

            if (isNeedAddSubLabel)
            {
                Controls.Add(new Literal { Text = "<table role='presentation' cellspacing='0' cellpadding='0'>" });
            }

            for (int i = 0; i < ChildControlCount; i++)
            {
                // create the sub label for each child control.
                if (isNeedAddSubLabel)
                {
                    Literal literal = new Literal();
                    string subLabel = LabelConvertUtil.GetTextByKey(ChildControlSubLabel, GetModuleName());
                    literal.Text = string.Format("<tr><td class='SubLabel ACA_Label font12px'>{0}</td><td class='SubControl'>", subLabel + (i + 1));
                    
                    Controls.Add(literal);
                }

                CreateChildControl(i, isNeedAddSubLabel);

                if (isNeedAddSubLabel)
                {
                    Controls.Add(new Literal { Text = "</td></tr>" });
                }
            }

            if (isNeedAddSubLabel)
            {
                Controls.Add(new Literal { Text = "</table>" });
            }

            ChildControlsCreated = true;
        }

        /// <summary>
        /// Raises the System.Web.UI.Control.PreRender event.
        /// </summary>
        /// <param name="e">A System.EventArgs that contains the event data.</param>
        protected override void OnPreRender(EventArgs e)
        {
            bool isAdmin = AccelaControlRender.IsAdminRender(this);

            if (!isAdmin && Page != null)
            {
                InitValidatorExtender();

                //Register the client scripts resource.
                ScriptManager.RegisterClientScriptResource(Page, GetType(), "Accela.Web.Controls.AccelaMultipleControl.js");

                string javascript = string.Format("AccelaMultipleControl.initializeControl('{0}', '{1}', '{2}');", ClientID, NextFocusControlID, ControlWidth);
                ScriptManager.RegisterStartupScript(Page, GetType(), "initMultipleCtl" + ClientID, javascript, true);
            }

            base.OnPreRender(e);
        }

        /// <summary>
        /// Displays the control on the client.
        /// </summary>
        /// <param name="writer">A System.Web.UI.HtmlTextWriter that contains the output stream to render on the client.</param>
        protected override void Render(HtmlTextWriter writer)
        {
            if (DesignMode || string.IsNullOrEmpty(ChildControlSubLabel))
            {
                base.Render(writer);
                return;
            }

            if (LayoutType == ControlLayoutType.Vertical)
            {
                writer.Write("<div class=\"MultipleControl_Vertical\">");
            }
            else
            {
                writer.Write("<div class=\"MultipleControl_Horizontal\">");
            }

            //Register a action to adjust the control's layout after document ready.
            string javascript = string.Format(
                                "AccelaMultipleControl.adjustSubFieldLayout('{0}', '{1}', '{2}');",
                                ClientID, 
                                Convert2HtmlType(ChildControlType),
                                LayoutType);

            ScriptManager.RegisterStartupScript(Page, GetType(), "adjustLayout" + ClientID, javascript, true);

            base.Render(writer);

            writer.Write("</div>");
        }

        #endregion Protected Control Event

        #region Private Methods

        /// <summary>
        /// Create the child control
        /// </summary>
        /// <param name="idSeqNbr">The id sequence number.</param>
        /// <param name="needAddSubLabel">Indicate whether need add sub label or not.</param>
        private void CreateChildControl(int idSeqNbr, bool needAddSubLabel)
        {
            string moduleName = GetModuleName();
            StringBuilder tooltip = new StringBuilder();

            // Add the parent label to tooltip.
            if (idSeqNbr == 0)
            {
                tooltip.Append(LabelConvertUtil.GetTextByKey(LabelKey, moduleName));
            }

            if (needAddSubLabel)
            {
                tooltip.Append(LabelConvertUtil.GetTextByKey(ChildControlSubLabelFullName, moduleName));
                tooltip.Append(idSeqNbr + 1);
            }

            if (IsRequired())
            {
                tooltip.Append(ACAConstant.BLANK);
                tooltip.Append(LabelConvertUtil.GetGlobalTextByKey("aca_required_field"));
            }

            // create the child control. Do NOT set the LabelKey, or it will set all label/help to all child controls.
            switch (ChildControlType)
            {
                case ControlType.Text:
                    AccelaTextBox textBox = new AccelaTextBox();
                    textBox.ID = string.Format("{0}{1}", CHILD_CONTROL_ID_PREFIX, idSeqNbr);
                    textBox.Width = TextBoxSet.Width.Value > 0 ? TextBoxSet.Width : Unit.Percentage(100);
                    textBox.MaxLength = TextBoxSet.MaxLength;
                    textBox.TrimValue = TextBoxSet.TrimValue;
                    TextBoxSet.LabelKey = LabelKey;

                    string waterMarkText = TextBoxSet.GetWatermarkText();

                    if (!string.IsNullOrEmpty(waterMarkText))
                    {
                        tooltip.Append(ACAConstant.BLANK);
                        tooltip.Append(waterMarkText);
                    }

                    textBox.WatermarkText = waterMarkText;
                    textBox.ToolTip = tooltip.ToString();

                    Controls.Add(textBox);

                    break;
                case ControlType.DropdownList:
                    AccelaDropDownList dropDownList = new AccelaDropDownList();
                    dropDownList.ID = string.Format("{0}{1}", CHILD_CONTROL_ID_PREFIX, idSeqNbr);
                    dropDownList.Width = DropDownListSet.Width.Value > 0 ? DropDownListSet.Width : Unit.Percentage(100);
                    dropDownList.ShowType = DropDownListSet.ShowType;
                    dropDownList.ToolTip = tooltip.ToString();

                    Controls.Add(dropDownList);

                    break;
            }
        }

        /// <summary>
        /// Initialize validator extender control
        /// </summary>
        private void InitValidatorExtender()
        {
            bool isRequired = IsRequired();
            bool isNeedDuplicateValidate = DuplicateValidate != null && DuplicateValidate.NeedValidate;

            if (!isRequired && !isNeedDuplicateValidate)
            {
                return;
            }

            // Create the custom validator controls for requried field.
            for (int i = 0; i < Children.Count; i++)
            {
                WebControl control = Children[i] as WebControl;

                if (control != null)
                {
                    if (isRequired)
                    {
                        string validateId = string.Format("RequiredValidate{0}", i);
                        CreateRequiredValidate(validateId, "AccelaMultipleControl.validatorForRequired", control);
                    }

                    if (isNeedDuplicateValidate)
                    {
                        string validateId = string.Format("DuplicateValidate{0}", i);
                        CreateDuplicateValidate(validateId, control.ID, control.ClientID);
                    }
                }
            }
        }

        /// <summary>
        /// Create duplicate validate control
        /// </summary>
        /// <param name="id">The new required validate control's id</param>
        /// <param name="controlToValidate">The input control id to validate</param>
        /// <param name="currentValidateClientID">The current child control when validate</param>
        private void CreateDuplicateValidate(string id, string controlToValidate, string currentValidateClientID)
        {
            var customValidator = new CustomValidator();
            string errorMsgLabelKey = DuplicateValidate != null ? DuplicateValidate.MessageLabelKey : string.Empty;
            string errorMessage = !string.IsNullOrEmpty(errorMsgLabelKey) ? LabelConvertUtil.GetTextByKey(errorMsgLabelKey, this) : string.Empty;

            customValidator.ID = id;
            customValidator.ControlToValidate = controlToValidate;
            customValidator.Display = ValidatorDisplay.None;
            customValidator.ClientValidationFunction = "AccelaMultipleControl.validatorForDuplicate";
            customValidator.ErrorMessage = errorMessage;
            customValidator.ValidateEmptyText = true;
            Controls.Add(customValidator);
            CreateValidatorCallbackExtender(customValidator.ID, currentValidateClientID);
        }

        /// <summary>
        /// Convert to html type.
        /// </summary>
        /// <param name="controlType">control type</param>
        /// <returns>html control type</returns>
        private string Convert2HtmlType(ControlType controlType)
        {
            string result = string.Empty;

            switch (controlType)
            {
                case ControlType.DropdownList:
                    result = "select";
                    break;

                case ControlType.Text:
                    result = "input[type=text]";
                    break;
            }

            return result;
        }

        #endregion Private Methods

        #region Inner Class

        /// <summary>
        /// This class provide the duplicate validate setting
        /// </summary>
        public class DuplicateValidateSetting
        {
            /// <summary>
            /// Gets or sets a value indicating whether need duplicate validate or not.
            /// </summary>
            public bool NeedValidate { get; set; }

            /// <summary>
            /// Gets or sets the duplicate validate message.
            /// </summary>
            public string MessageLabelKey { get; set; }
        }

        #endregion Inner Class
    }
}
